onnxruntime/contrib_ops/cuda/bert/attention_impl.cu

```diff
@@ -0,0 +1,424 @@
+template <typename T, unsigned TPB>
+__device__ inline void scaledSoftmaxSmall(
+    const int ld, const int lastValid, const float rsqrtHeadSize, const T* input, T* output) {
+  using BlockReduce = cub::BlockReduce<float, TPB>;
+
+  __shared__ typename BlockReduce::TempStorage tmpStorage;
+
+  __shared__ float rZ;
+
+  const int offset = (blockIdx.y * gridDim.x + blockIdx.x) * ld;
+
+  const float w(rsqrtHeadSize);
+  cub::Sum sum;
+  float threadData(0);
+
+  const int idx = offset + threadIdx.x;
+  if (threadIdx.x < lastValid) {
+    const float val = input[idx];
+    threadData = exp(val * w);
+  }
+
+  const auto Z = BlockReduce(tmpStorage).Reduce(threadData, sum);
+
+  if (threadIdx.x == 0) {
+    rZ = (1.f) / Z;
+  }
+  __syncthreads();
+
+  if (threadIdx.x < ld) {
+    // this will be 0 for threadIdx.x >= lastValid
+    output[idx] = T(threadData * rZ);
+  }
+}
+
+template <typename T, unsigned TPB>
+__global__ void maskedScaledSoftmaxKernelSmall(
+    const int ld, const float rsqrtHeadSize, const int* maskIdx, const T* input, T* output) {
+  __shared__ int lastValid;
+
+  if (threadIdx.x == 0) {
+    lastValid = min(ld, maskIdx[blockIdx.y]);
+  }
+  __syncthreads();
+
+  scaledSoftmaxSmall<T, TPB>(ld, lastValid, rsqrtHeadSize, input, output);
+}
+
+template <typename T, unsigned TPB>
+__global__ void maskedScaledSoftmaxKernel(
+    const int ld, const float rsqrtHeadSize, const int* maskIdx, const T* input, T* output) {
+  __shared__ int lastValid;
+
+  if (threadIdx.x == 0) {
+    lastValid = min(ld, maskIdx[blockIdx.y]);
+  }
+  __syncthreads();
+  scaledSoftmax<T, TPB>(ld, lastValid, rsqrtHeadSize, input, output);
+}
+
+template <typename T>
+int computeMaskedScaledSoftmax(cudaStream_t stream, const int ld, const int B, const int N, const float rsqrtHeadSize,
+                               const int* maskIdx, const T* input, T* output) {
+  // Mask idx is of length B and assumes the valid region is contiguous starting
+  // from the beginning of the sequence
+
+  const dim3 grid(ld * N, B, 1);
+
+  if (ld <= 32) {
+    const int blockSize = 32;
+    maskedScaledSoftmaxKernelSmall<T, blockSize>
+        <<<grid, blockSize, 0, stream>>>(ld, rsqrtHeadSize, maskIdx, input, output);
+  } else if (ld <= 128) {
+    const int blockSize = 128;
+    maskedScaledSoftmaxKernelSmall<T, blockSize>
+        <<<grid, blockSize, 0, stream>>>(ld, rsqrtHeadSize, maskIdx, input, output);
+  } else if (ld == 384) {
+    const int blockSize = 384;
+    maskedScaledSoftmaxKernelSmall<T, blockSize>
+        <<<grid, blockSize, 0, stream>>>(ld, rsqrtHeadSize, maskIdx, input, output);
+  } else {
+    const int blockSize = 256;
+    maskedScaledSoftmaxKernel<T, blockSize>
+        <<<grid, blockSize, 0, stream>>>(ld, rsqrtHeadSize, maskIdx, input, output);
+  }
+
+  CUDA_CALL(cudaPeekAtLastError());
+  return 0;
+}
+
+
+template <typename T>
+cublasStatus_t inline cublasGemm(cublasHandle_t handle, cublasOperation_t transa, cublasOperation_t transb, int m,
```

Comment:
>cublasGemm [](start = 22, length = 10)

Seems dup of cublasGemmHelper: [https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h#L21](https://github.com/microsoft/onnxruntime/blob/master/onnxruntime/core/providers/cuda/shared_inc/fpgeneric.h#L21)

Reply:
Right. I'll remove them.